Dbeaver/pro#4149 add navigator asc sorting

labanovichttps opened the pull request:
<img width="481" alt="image" src="https://github.com/user-attachments/assets/fed9c3e7-30d6-4258-8f99-f5d73ef20e9d" />

XLSX For testing
[testmasha.xlsx](https://github.com/user-attachments/files/19647073/testmasha.xlsx)

labanovichttps committed:
dbeaver/pro#4216 Add AI object describe

labanovichttps committed:
Merge remote-tracking branch 'origin/devel' into devel

labanovichttps committed:
Merge remote-tracking branch 'origin/devel' into devel

labanovichttps committed:
Merge remote-tracking branch 'origin/devel' into devel

labanovichttps committed:
dbeaver/pro#4149 Add general navigator sorting

E1izabeth commented on plugins/org.jkiss.dbeaver.model.ai/src/org/jkiss/dbeaver/model/ai/AIAssistant.java:
```diff
@@ -59,4 +59,14 @@ CommandResult command(
+     *  Generates the description for DBSObject.
+     */
+    @NotNull
+    String describe(
```
Please remove irrelevant changes

E1izabeth reviewed (commented):
We sort all nodes twice now under some conditions. Isn't it better to write one comparator?
Also there is a setting `ModelPreferences.NAVIGATOR_SORT_ALPHABETICALLY`. Shouldn't alphabetical sorting depend on it?

labanovichttps committed:
dbeaver/pro#4149 Resolve git issues

labanovichttps commented:
> We sort all nodes twice now under some conditions. Isn't it better to write one comparator? Also there is a setting `ModelPreferences.NAVIGATOR_SORT_ALPHABETICALLY`. Shouldn't alphabetical sorting depend on it?

Added sorting for `ModelPreferences.NAVIGATOR_SORT_ALPHABETICALLY.` 
I also added a check for only one sorting instead of twice.

labanovichttps committed:
dbeaver/pro#4149 Edit const modifier

E1izabeth commented on plugins/org.jkiss.dbeaver.model/src/org/jkiss/dbeaver/model/navigator/DBNUtils.java:
```diff
@@ -48,6 +46,8 @@
 public class DBNUtils {
 
     private static final Log log = Log.getLog(DBNUtils.class);
+    private static final Comparator<DBNNode> ALPHANUMERIC_COMPARATOR_COMPARATOR = (o1, o2) -> AlphanumericComparator.getInstance()
```
Maybe just `ALPHANUMERIC_COMPARATOR`?
```suggestion
    private static final Comparator<DBNNode> ALPHANUMERIC_COMPARATOR = (o1, o2) -> AlphanumericComparator.getInstance()
```

labanovichttps replied:
fixed

ShadelessFox replied:
`Comparator.comparing(DBNNode::getNodeDisplayName, AlphanumericComparator.getInstance())`

E1izabeth reviewed (commented):
It would be nice to write a couple of unit tests for this logic

labanovichttps committed:
dbeaver/pro#4149 Fix comparator name

ShadelessFox commented on plugins/org.jkiss.dbeaver.model/src/org/jkiss/dbeaver/model/navigator/DBNUtils.java:
```diff
@@ -135,14 +134,17 @@ private static void sortNodes(DBNNode[] children)
                 } else if (prefStore.getBoolean(ModelPreferences.NAVIGATOR_SORT_ALPHABETICALLY) || isMergedEntity(firstChild)) {
                     if (!(firstChild instanceof DBNContainer)) {
-                        Arrays.sort(children, NodeNameComparator.INSTANCE);
+                        Arrays.sort(children, ALPHANUMERIC_COMPARATOR_COMPARATOR);
```
`NodeNameComparator` is not used anymore; please delete it

ShadelessFox commented on plugins/org.jkiss.dbeaver.model/src/org/jkiss/dbeaver/model/navigator/DBNUtils.java:
```diff
@@ -135,14 +134,17 @@ private static void sortNodes(DBNNode[] children)
+                        Arrays.sort(children, NodeFolderComparator.INSTANCE);
+                    } else if (!(firstChild instanceof DBNContainer)) {
+                        Arrays.sort(children, ALPHANUMERIC_COMPARATOR_COMPARATOR);
+                    }
```
I don't quite understand what this change does. Can you please elaborate?

labanovichttps replied:
For example, opening a connection and seeing Databases/Administrator, etc. It applies NodeFolderComparator.INSTANCE as default, but if you open any schema or table it applies ALPHANUMERIC_COMPARATOR_COMPARATOR for comparing like this 
![image](https://github.com/user-attachments/assets/77d4e6d0-b0f2-4dbf-834c-c253570ee693)

labanovichttps committed:
dbeaver/pro#4149 Fix comparator

labanovichttps committed:
dbeaver/pro#4149 Refactor comparing elements

Matvey16 committed:
Merge branch 'devel' into dbeaver/pro#4149-add-navigator-asc-sorting

Matvey16 commented:
Verified
